### PR TITLE
improve threading correctness in URLSession

### PR DESCRIPTION
**Pull request by weissi**
this is really only the beginning to prevent everything from breaking with the simplest test that does use more than one thread. Thread-safely seems to be large absent in this part of foundation 😢. 
Unfortunately, I seem to be going down a rabbit hole here but it's a start.

### CODE REVIEW

File: Foundation/URLSession/URLSessionTask.swift [hunk 1]
```diff
@@ -555,7 +517,9 @@ extension _ProtocolClient : URLProtocolClient {
             session.delegateQueue.addOperation {
                 delegate.urlSession(session, task: task, didCompleteWithError: nil)
                 task.state = .completed
-                session.taskRegistry.remove(task)
+                task.workQueue.async {
```
Comment (pushkarnk) [inline Foundation/URLSession/URLSessionTask.swift]: I think this will fix an extremely intermittent crash a few of us have seen.

**Comment by pushkarnk** [comment]
@swift-ci please test

File: Foundation/URLSession/URLSessionTask.swift [hunk 2]
```diff
@@ -62,7 +59,8 @@ open class URLSessionTask : NSObject, NSCopying {
     }
     internal init(session: URLSession, request: URLRequest, taskIdentifier: Int, body: _Body) {
         self.session = session
-        self.workQueue = session.workQueue
+        /* make sure we're actually having a serial queue as it's used for synchronization */
+        self.workQueue = DispatchQueue.init(label: "org.swift.URLSessionTask.WorkQueue", target: session.workQueue)
```
Comment (pushkarnk) [inline Foundation/URLSession/URLSessionTask.swift]: Makes sense. However, I recall we had replaced the use of DispatchQueues with semaphores in attempts to work around a weird and intermittent hang in `DispatchQueue.async` - something we could reproduce only in a couple of local environments.

File: Foundation/URLSession/URLSession.swift [hunk 3]
```diff
@@ -171,12 +171,13 @@
 import CoreFoundation
 import Dispatch
 
-
+fileprivate let globalVarSyncQ = DispatchQueue(label: "org.swift.Foundation.URLSession.GlobalVarSyncQ")
```
Comment (pushkarnk) [inline Foundation/URLSession/URLSession.swift]: Thanks! This was long pending and ignored :(

**Comment by weissi** [comment]
@swift-ci please test

**Comment by weissi** [comment]
@swift-ci please test

File: Foundation/URLSession/URLSessionTask.swift (hunk 2 above)
Comment (weissi) [inline Foundation/URLSession/URLSessionTask.swift]: @pushkarnk well, if that's the case I want to know and I'll fix the issue.

**Comment by weissi** [comment]
tests all passed and approved, so merging.

**Pull request merged**